### ISSUE TITLE
How to blacklist token?

### ISSUE THREAD

**Issue by gandra** [fastify/fastify-jwt#56]
Hi,
when user `logout` I would like to `blacklist` a token. Is there a way to do it?

**Comment by cemremengu** [comment]
I think that is an external concern/logic that should be implemented separately. You will basically check if it blacklisted before verifying the token.

**Comment by zuohuadong** [comment]
> I think that is an external concern/logic that should be implemented separately. You will basically check if it blacklisted before verifying the token in `onRequest` hook and blacklist it when logout endpoint is called.

https://github.com/auth0/express-jwt

### Revoked tokens
It is possible that some tokens will need to be revoked so they cannot be used any longer. You can provide a function as the `isRevoked` option. The signature of the function is `function(req, payload, done)`:
* `req` (`Object`) - The express `request` object.
* `payload` (`Object`) - An object with the JWT claims.
* `done` (`Function`) - A function with signature `function(err, revoked)` to be invoked once the check to see if the token is revoked or not is complete.
  * `err` (`Any`) - The error that occurred.
  * `revoked` (`Boolean`) - `true` if the JWT is revoked, `false` otherwise.

For example, if the `(iss, jti)` claim pair is used to identify a JWT:
```javascript
var jwt = require('express-jwt');
var data = require('./data');
var utilities = require('./utilities');

var isRevokedCallback = function(req, payload, done){
  var issuer = payload.iss;
  var tokenId = payload.jti;

  data.getRevokedToken(issuer, tokenId, function(err, token){
    if (err) { return done(err); }
    return done(null, !!token);
  });
};

app.get('/protected',
  jwt({secret: 'shhhhhhared-secret',
    isRevoked: isRevokedCallback}),
  function(req, res) {
    if (!req.user.admin) return res.sendStatus(401);
    res.sendStatus(200);
  });
```

**Comment by jsumners** [comment]
@zuohuadong what is your point in providing the documentation for a completely different plugin?

**Comment by zuohuadong** [comment]
> @zuohuadong what is your point in providing the documentation for a completely different plugin?


It can be implemented and has a good reference.
Fastify should be more advanced than express, so it should have a similar implementation.

**Comment by jsumners** [comment]
This plugin seems to provide everything necessary when I read the documentation. One simply has to add a `preValidation` method that checks if the decoded/verified token has data that is not allowed:

```js
// register fastify-jwt first
fastify.addHook('preValidation', async function (request, reply) {
  try {
    await request.jwtVerify()
  } catch (err) { return reply.send(err) }

  if (request.user === 'bannedUser') return reply.code(401).send(Error('Unauthorized'))
})
```

**Comment by mcollina** [comment]
I think they are asking how to do the same thing with fastify-jwt. Would you like to senf a PR?

**Comment by zuohuadong** [comment]
> This plugin seems to provide everything necessary when I read the documentation. One simply has to add a `preValidation` method that checks if the decoded/verified token has data that is not allowed:
> 
> ```js
> // register fastify-jwt first
> fastify.addHook('preValidation', async function (request, reply) {
>   try {
>     await request.jwtVerify()
>   } catch (err) { return reply.send(err) }
> 
>   if (request.user === 'bannedUser') return reply.code(401).send(Error('Unauthorized'))
> })
> ```


Yes, I hope that fastify-jwt can also have such an implementation.
I prefer fastify.

**Comment by francisbrito** [comment]
```js
module.exports = async function (fastify, options) {
  fastify.register(require('fastify-jwt'), {
    secret: 'foobar',
  })

  fastify.addHook('onRequest', async (request, reply) => {
    try {
      await request.jwtVerify()
    } catch (err) {
      reply.send(err)
    }
  })

  fastify.get('/', async (request, reply) => {
    const [, token] = request.headers.authorization.split(' ')
    const { jti, sub: userId, email } = fastify.jwt.decode(token) // NOTE: decoding is safe since we already verified the token before hitting this route.

    const tokenWasBanned = await checkIfTokenWasBanned(jti)

    if (tokenWasBanned) {
      reply.status(401).send({ code: 'E_TOKEN_WAS_BANNED', message: 'Please authenticate again.' })
      return
    }

    const userWasBanned = await checkIfUserWasBannedByIdOrEmail({ userId, email })

    if (userWasBanned) {
      reply.status(401).send({ code: 'E_USER_WAS_BANNED', message: 'User was banned. Please contact support.' })
      return
    }

    reply.send({ code: 'OK', message: 'it works!' })
  })
}

// ideally these functions would do a query against some sort of storage to determine their outcome
async function checkIfTokenWasBanned(tokenId) {
  return Promise.resolve(false)
}
async function checkIfUserWasBannedByIdOrEmail({ userId, email }) {
  return Promise.resolve(false)
}
```

@zuohuadong do you think an example like this works for you?
@Eomm is it ok if I open a PR with an example on how to setup token black-listing? (maybe even using Redis as storage)

**Comment by Eomm** [comment]
I was thinking this issue more like a feature:

```js
fastify.register(require('fastify-jwt'), {
  secret: 'supersecret',
  untrusted: (request, decoded) => {
      if(decoded.sub.email === 'banned@email.com'){
          throw new Error('Banned email')
      }
      // we could manage promise also
  }
})
```

And, when the user set an `untrusted` function it will be evaluated when `.jwtVerify()` is called.

But an example could work as well 👍
Consider to:
- create an `example/` folder
- wrap your code in a `onRequest` hook to improve reusability

**Comment by francisbrito** [comment]
I think adding it as a feature makes sense. I'll proceed with the PR

**Comment by lludol** [comment]
Hey guys, what are the updates?
Do you need help for the PR?

I need this feature right now 😁

**Comment by mcollina** [comment]
go ahead!

Il giorno lun 4 nov 2019 alle ore 15:20 Ludovic Lerus <
notifications@github.com> ha scritto:

**Comment by Eomm** [comment]
There is this PR https://github.com/fastify/fastify-jwt/pull/71 that is quite finished, we need to wait @francisbrito 💪

**Comment by francisbrito** [comment]
@lludol @Eomm sorry for the delay (had little time for OSS in the past weeks). I have updated the PR with the requested changes. Please let me know if you have any additional suggestion regarding it.

**Comment by Eomm** [comment]
Released in v1.2.0

Thank you @francisbrito for your contribution! 💪

**Comment by onosendi** [comment]
>NOTE: decoding is safe since we already verified the token before hitting this route.

@francisbrito in what situation would it be unsafe to decode a token before verifying?
